### PR TITLE
[optimizer] Make optimizer independent of weight

### PR DESCRIPTION
**Pull request by kparichay**
Cleanup optimizer interface where some extra interfaces have been
removed, and arguments interface has been updated.
Further, the dependency on weight header has been removed, and added an
optimizer context interface which provides an interface to weight
related tensors required by the optimizer.

Update the interface for the layer.h setProperty.

Signed-off-by: Parichay Kapoor <pk.kapoor@samsung.com>

### CODE REVIEW

File: nntrainer/optimizers/optimizer_devel.h [hunk 1]
```diff
@@ -88,25 +85,10 @@ class Optimizer : public ml::train::Optimizer {
...
-  virtual void setProperty(const std::string &key,
-                           const std::string &value) = 0;
-
   /**
    * @brief     initialize optimizer.
-   * @retval #ML_ERROR_NONE Successful.
-   * @retval #ML_ERROR_INVALID_PARAMETER invalid parameter.
    */
-  virtual int initialize() = 0;
+  virtual void initialize(){};
```
Comment (zhoonit) [inline nntrainer/optimizers/optimizer_devel.h:91]: As we are making a api breaking change, may be finalize to keep modules consistent?

**Review by zhoonit — APPROVED** [review]
LGTM with a off-the-topic suggestion :)

File: nntrainer/optimizers/optimizer_devel.h (hunk 1 above)
Comment (kparichay) [reply nntrainer/optimizers/optimizer_devel.h:91]: will update.

Comment (kparichay) [reply nntrainer/optimizers/optimizer_devel.h]: done.

**Review by jijoongmoon — COMMENTED** [review]
LGTM

**Review by jijoongmoon — APPROVED** [review]

**Pull request merged**